pkg/config/ciphersuites.go

```diff
@@ -0,0 +1,81 @@
+// Copyright Project Contour Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package config
+
+import (
+	"fmt"
+	"strings"
+)
+
+// TLSCiphers holds a list of TLS ciphers
+type TLSCiphers []string
+
+// Validate ciphers. Returns error on unsupported cipher.
+func (tlsCiphers TLSCiphers) Validate() error {
+	invalidCiphers := []string{}
+	for _, cipher := range tlsCiphers {
+		trimmed := strings.TrimSpace(cipher)
+		if _, ok := ValidTLSCiphers[trimmed]; !ok {
+			invalidCiphers = append(invalidCiphers, trimmed)
+		}
+	}
+	if len(invalidCiphers) > 0 {
+		return fmt.Errorf("invalid ciphers: %s", strings.Join(invalidCiphers, ","))
+	}
+	return nil
+}
+
+// DefaultTLSCiphers contains the list of default ciphers used by Contour. A handful are
+// commented out, as they're arguably less secure. They're also unnecessary
+// - most of the clients that might need to use the commented ciphers are
+// unable to connect without TLS 1.0, which contour never enables.
+//
+// This list is ignored if the client and server negotiate TLS 1.3.
+//
+// The commented ciphers are left in place to simplify updating this list for future
+// versions of envoy.
+var DefaultTLSCiphers = TLSCiphers([]string{
+	"[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]",
+	"[ECDHE-RSA-AES128-GCM-SHA256|ECDHE-RSA-CHACHA20-POLY1305]",
+	"ECDHE-ECDSA-AES128-SHA",
+	"ECDHE-RSA-AES128-SHA",
+	//"AES128-GCM-SHA256",
+	//"AES128-SHA",
+	"ECDHE-ECDSA-AES256-GCM-SHA384",
+	"ECDHE-RSA-AES256-GCM-SHA384",
+	"ECDHE-ECDSA-AES256-SHA",
+	"ECDHE-RSA-AES256-SHA",
+	//"AES256-GCM-SHA384",
+	//"AES256-SHA",
+})
+
+// ValidTLSCiphers contains the list of TLS ciphers that Envoy supports
+// See: https://www.envoyproxy.io/docs/envoy/latest/api-v3/extensions/transport_sockets/tls/v3/common.proto#extensions-transport-sockets-tls-v3-tlsparameters
+// Note: This list is a superset of what is valid for stock Envoy builds and those using BoringSSL FIPS.
```

Comment:
what happens if a user configures a FIPS cipher suite as allowed, but they're not using a FIPS Envoy build? It sounds like from your previous comment that Envoy would just ignore it? If so, I think that's probably fine; it's a reasonable tradeoff for enabling this to work for FIPS environments too.

Reply:
The Envoy docs on the allowed ciphers are a little misleading, the literal strings of the two ciphers that "differ" for the FIPS build `ECDHE-ECDSA-AES128-GCM-SHA256` and `ECDHE-RSA-AES128-GCM-SHA256` are accepted by a plain Envoy and not ignored even though the docs make it look like Envoy wouldn't accept that string (though of course if you see `[ECDHE-ECDSA-AES128-GCM-SHA256|ECDHE-ECDSA-CHACHA20-POLY1305]` it makes sense).

Reply:
Ah reading https://commondatastorage.googleapis.com/chromium-boringssl-docs/ssl.h.html#Cipher-suite-configuration helps explain that notation, I wasn't aware of what it meant.

Reply:
```
If configuring a server, one may also configure equal-preference groups to partially respect the client's preferences when SSL_OP_CIPHER_SERVER_PREFERENCE is enabled. Ciphers in an equal-preference group have equal priority and use the client order. This may be used to enforce that AEADs are preferred but select AES-GCM vs. ChaCha20-Poly1305 based on client preferences. An equal-preference is specified with square brackets, combining multiple selectors separated by |. For example:

[TLS_ECDHE_ECDSA_WITH_CHACHA20_POLY1305_SHA256|TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256]

Once an equal-preference group is used, future directives must be opcode-less. Inside an equal-preference group, spaces are not allowed.
```

Reply:
yep exactly so we allow one to specify the preference groups *or* the standalone cipher that is FIPS compatible that Envoy will accept